### PR TITLE
Also use a specialized implementation for wset. For now only non enum…

### PR DESCRIPTION
**Pull request by baldersheim**
… versions.
@geirst PR

### CODE REVIEW

**Comment by baldersheim** [comment]
@geirst Ping

File: searchlib/src/vespa/searchlib/features/dotproductfeature.cpp [hunk 1]
```diff
@@ -542,29 +596,55 @@ FeatureExecutor * createTypedArrayExecutor(const IAttributeVector * attribute,
...
-    if (attribute->isStringType()) {
-        if (attribute->hasEnum()) {
-            dotproduct::wset::EnumVector vector(attribute);
-            WeightedSetParser::parse(prop.get(), vector);
-            return &stash.create<dotproduct::wset::DotProductExecutor<dotproduct::wset::EnumVector, WeightedEnumContent>>(attribute, vector);
-        } else {
-            dotproduct::wset::StringVector vector;
-            WeightedSetParser::parse(prop.get(), vector);
-            return &stash.create<dotproduct::wset::DotProductExecutor<dotproduct::wset::StringVector, WeightedConstCharContent>>(attribute, vector);
+template <typename A, typename V>
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/features/dotproductfeature.cpp]: Consider more descriptive type names.

**Review by geirst — APPROVED** [review]
Overall looks good, some comments regarding naming to consider.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.h [hunk 2]
```diff
@@ -55,71 +55,106 @@ class VectorBase {
...
 
+/**
+ * Common base for handling execution for all wset dot product executors.
+ * Only cares about the underlying value type, not the concrete type of the
+ * attribute vector itself.
+ */
+template <typename BaseType>
+class DotProductExecutorBase : public fef::FeatureExecutor {
+public:
+    using AT = multivalue::WeightedValue<BaseType>;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/dotproductfeature.h]: I kept the naming the same for the array and the wset version. I will leave that to a later refactoring.

File: searchlib/src/vespa/searchlib/features/dotproductfeature.h [hunk 3]
```diff
@@ -55,71 +55,106 @@ class VectorBase {
...
     }
 };
 
+/**
+ * Common base for handling execution for all wset dot product executors.
+ * Only cares about the underlying value type, not the concrete type of the
+ * attribute vector itself.
+ */
+template <typename BaseType>
+class DotProductExecutorBase : public fef::FeatureExecutor {
```
[the remark this answers is not in the document]
Comment (geirst) [reply searchlib/src/vespa/searchlib/features/dotproductfeature.h]: Ok

**Pull request merged**